csrc/abstract_tensor.h

```diff
@@ -338,4 +351,120 @@ struct AbstractTensor {
   AbstractTensor& strip();
 };
 
+//! This is a wrapper around AbstractTensor which propagates a set of tags for
+//! each axis. The tags can be any hashable type with equality; typically Tag
+//! would be an enum. See AbstractMatmulTensor, in which case Tag =
+//! MatmulDimRole.
+template <typename Tag>
+struct TaggedAbstractTensor : AbstractTensor {
+  //! Holds a set of tags for each dimension
+  std::vector<std::unordered_set<Tag>> tags;
+
+  const std::unordered_set<Tag>& getTags(int64_t i) const {
+    i = wrapDim(i, (int64_t)domain.size());
+    return tags[i];
+  }
+
+  bool hasTag(int64_t i, Tag tag) const {
+    return (bool)getTags(i).count(tag);
+  }
+
+  //! Return tag if there is a single tag, otherwise nullopt
+  //!
+  //! This is just a convenience function for the common case where axes with
+  //! different tags have not been merged. In these cases there is a single Tag
+  //! for each axis and it is cumbersome to extract it manually each time it's
+  //! needed.
+  std::optional<Tag> getTag(int64_t i) const {
+    const auto& tags = getTags(i);
+    if (tags.size() == 1) {
+      return *tags.begin();
+    }
+    return std::nullopt;
+  }
+
+  template <typename T>
+  bool operator==(T&& t) const {
+    if constexpr (std::is_same_v<TaggedAbstractTensor<Tag>, std::decay_t<T>>) {
+      return domain == t.domain && tags == t.tags;
+    } else {
+      // Only compare true for other TaggedAbstractTensors
+      return false;
+    }
+  }
+
+  using AbstractTensor::split;
+
+  AbstractTensor& split(int64_t axis, Val* factor, bool inner_split = true)
+      override {
+    axis = wrapDim(axis, (int64_t)domain.size());
+    // copy tags from original axis
+    tags.insert(tags.begin() + axis + 1, tags.at(axis));
+    return AbstractTensor::split(axis, factor, inner_split);
+  }
```

Comment:
Looks like this class has a lot of code duplication with `AbstractTensor`. Does it make sense to refactor the code like below?

```C++
template <typename Info>
struct AbstractTensorWithInfo {
  std::vector<std::pair<AbstractId, Info>> domain;
  ...
  AbstractTensor& AbstractTensor::merge(int64_t axis_o, int64_t axis_i) {
    wrap dim
    auto [id_o, info_o] = ...;
    auto [id_i, info_i] = ...;
    AbstractId result_id = do_the_merge(id_o, id_i);
    Info result_info = Info::merge(info_o, info_i);
    auto result = std::make_pair(result_id, result_info);
    // write result to outer
  }
  ...
};

struct NoInfo {
  static NoInfo merge(NoInfo, NoInfo) { return {}; }
  ...
} ; 

using AbstractTensor = AbstractTensorWithInfo<NoInfo>;

template <typename Tag>
struct TagInfo {
  std::unordered_set<Tag> tags;
  static TagInfo merge(TagInfo, TagInfo) { return ...; }
  ...
};

template <typename Tag>
using TaggedAbstractTensor = AbstractTensorWithInfo<TagInfo<Tag>>;
```